### PR TITLE
[tests-only][full-ci]Adds test related to renaming of file

### PR DESCRIPTION
**Pull request by amrita-shrestha**
## Description
This PR adds fixes for unexpected behavior on test related to file rename 

## Related Issue
Adds acceptance test coverage for: https://github.com/owncloud/web/issues/4893
Note: that issue has been fixed, and this test covers the now-good behavior.

## How Has This Been Tested?
- Locally
- CI

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [x] Tests

## Checklist:
- [ ] Code changes
- [ ] Unit tests added
- [x] Acceptance tests added
- [ ] Documentation ticket raised: <link>

### CODE REVIEW

**Comment by phil-davis** [comment]
@amrita-shrestha you need to click the button to "sign" the CLA.

File: tests/acceptance/features/webUIFavouritesActionMenu/favouritesActionMenu.feature [hunk 1]
```diff
@@ -0,0 +1,16 @@
+Feature: Visibilty of the file or folder actions menu options on favourites folder
+    As a user
+    I want to see the actions menu options of the file or folder on favourites folder
+    So that the menu options of the selected file type or folder are visible to me
+
```
Comment (Talank) [inline tests/acceptance/features/webUIFavouritesActionMenu/favouritesActionMenu.feature:5]: we put 2 empty lines before each scenarios. like you did in `tests/acceptance/features/webUIBatchAction/batchAction.feature`

Comment (phil-davis) [reply tests/acceptance/features/webUIFavouritesActionMenu/favouritesActionMenu.feature:5]: Note: we do that so that we can tag scenarios in future, and still have an empty line between the previous scenario and the tags of the next scenario. And we can do that without line-numbers changing (that might be referenced in expected-failures files or...)

File: tests/acceptance/features/webUIRenameFiles/renameFilesWithoutReload.feature [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+  I want to rename files
+  So that I can organise my data structure
+
+  Background:
+    Given user "Alice" has been created with default attributes and without skeleton files
+    And user "Alice" has created file "index.jpeg"
+    And user "Alice" has logged in using the webUI
+
+
+  Scenario: Rename file extension with reload
```
Comment (SwikritiT) [inline tests/acceptance/features/webUIRenameFiles/renameFilesWithoutReload.feature:12]: This scenario can be added to this file https://github.com/owncloud/web/blob/master/tests/acceptance/features/webUIRenameFiles/renameFiles.feature , you don't need to make a separate file for that. Maybe you can make this a scenario outline and test different formats too?

File: tests/acceptance/features/webUIRenameFiles/renameFilesWithoutReload.feature [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+    When the user tries to rename file "index.jpeg" to "index.jpg"
+    Then file "index.jpg" should be listed as "inde.jpeg" without reload
+    And file "index.jpg" should be listed on the sidebar
```
Comment (SwikritiT) [inline tests/acceptance/features/webUIRenameFiles/renameFilesWithoutReload.feature:15]:
```suggestion
    Then file "inde.jpeg" should be listed on the webUI
    When the user reloads the current page of the webUI
    Then file "index.jpg" should be listed on the webUI
```

File: tests/acceptance/features/webUIBatchAction/batchAction.feature [hunk 4]
```diff
@@ -0,0 +1,65 @@
+Feature: Visibilty of the batch actions menu
+  As a user
+  I want to see the batch action menu options
+  So that the action menu options of the selected file type or folder are visible to me
+
+
+  Scenario: View batch action menu for a file/folder on shared with me
```
Comment (SwikritiT) [inline tests/acceptance/features/webUIBatchAction/batchAction.feature:7]:
```suggestion
  Scenario: View batch action menu for a file/folder on shared with me  page
```

File: tests/acceptance/features/webUIBatchAction/batchAction.feature [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+      | Brian    |
+    And user "Alice" has created folder "simple-folder"
+    And user "Alice" has shared folder "simple-folder" with user "Brian"
+    And user "Brian" has logged in using the webUI
+    And the user has browsed to the shared-with-me page
+    When the user select the folder "simple-folder" using the webUI
+    Then the decline button should be displayed on shared-with-me page
+
+
+  Scenario: View batch action menu for a file/folder on shared with others
```
Comment (SwikritiT) [inline tests/acceptance/features/webUIBatchAction/batchAction.feature:20]:
```suggestion
  Scenario: View batch action menu for a file/folder on the shared with others page
```

File: tests/acceptance/features/webUIBatchAction/batchAction.feature [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+  Scenario: View batch action menu for a file/folder on shared with me 
+    Given these users have been created with default attributes and without skeleton files:
+      | username |
+      | Alice    |
+      | Brian    |
+    And user "Alice" has created folder "simple-folder"
+    And user "Alice" has shared folder "simple-folder" with user "Brian"
+    And user "Brian" has logged in using the webUI
+    And the user has browsed to the shared-with-me page
+    When the user select the folder "simple-folder" using the webUI
```
Comment (SwikritiT) [inline tests/acceptance/features/webUIBatchAction/batchAction.feature:16]:
```suggestion
    When the user selects the folder "simple-folder" using the webUI
```

File: tests/acceptance/features/webUIBatchAction/batchAction.feature [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+      | Brian    |
+    And user "Brian" has created folder "brain-folder"
+    And user "Brian" has shared folder "brain-folder" with user "Alice"
+    And user "Brian" has logged in using the webUI
+    And the user has browsed to the shared-with-others page
+    When the user select the folder "brain-folder" using the webUI
+    Then the delete button should be displayed on shared-with-others page
+
+
+  Scenario: View action menu for a file/folder on shared with me without auto_accept_share
```
Comment (SwikritiT) [inline tests/acceptance/features/webUIBatchAction/batchAction.feature:33]:
```suggestion
  Scenario: View action menu for a file/folder on the shared with me page without auto_accept_share
```

File: tests/acceptance/features/webUIBatchAction/batchAction.feature [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    And user "Brian" has logged in using the webUI
+    And the user has browsed to the shared-with-me page
+    When the user select the folder "simple-folder" using the webUI
+    Then the following batch action buttons should be displayed on shared page
+      | buttons |
+      | Accept  |
+      | Decline |
+
+
+  Scenario: View action menu for a file/folder on trashbin
```
Comment (SwikritiT) [inline tests/acceptance/features/webUIBatchAction/batchAction.feature:53]:
```suggestion
  Scenario: View action menu for a file/folder on the trashbin page
```

**Review by sakshamgurung — COMMENTED** [review]
@amrita-shrestha in my opinion you can collectively make changes and push. No need to push rapidly.

File: tests/acceptance/stepDefinitions/favouritesActionMenuContext.js [hunk 9]
```diff
@@ -0,0 +1,15 @@
+const { client } = require('nightwatch-api')
+const { Then, When } = require('@cucumber/cucumber')
+
+When('the user select the file/folder {string} using the webUI', function (fileName) {
```
Comment (SwikritiT) [inline tests/acceptance/stepDefinitions/favouritesActionMenuContext.js:4]:
```suggestion
When('the user selects the file/folder {string} using the webUI', function (fileName) {
```

File: tests/acceptance/stepDefinitions/favouritesActionMenuContext.js [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+  return client.page.FilesPageElement.filesList().toggleFileOrFolderCheckbox('enable', fileName)
+})
+
+Then('the batch action button should not be displayed', function () {
```
Comment (SwikritiT) [inline tests/acceptance/stepDefinitions/favouritesActionMenuContext.js:8]:
```suggestion
Then('the batch action button should not be visible on the webUI', function () {
```

File: tests/acceptance/stepDefinitions/filesContext.js [hunk 11]
```diff
@@ -697,6 +697,27 @@ Then(
   }
 )
 
+When('the user tries to rename file {string} to {string}', function (fromName, toName) {
```
Comment (SwikritiT) [inline tests/acceptance/stepDefinitions/filesContext.js:700]: We have similar step here https://github.com/owncloud/web/blob/f9779f30d472aebfc5d7e4755d749f99e4f9a4f6/tests/acceptance/stepDefinitions/filesContext.js#L344 
if this rename is through some other way please add something like `using the context menu`

File: tests/acceptance/stepDefinitions/filesContext.js [hunk 12]
```diff
@@ -697,6 +697,27 @@ Then(
...
+  'file {string} should be listed as {string} without reload',
+  async function (expectedFileName, actualFileName) {
+    const fileRename = await client.page.FilesPageElement.filesList().waitForFileRenameVisible(
+      expectedFileName,
+      'file'
+    )
+    return assert.strictEqual(fileRename, actualFileName)
+  }
+)
+
```
Comment (SwikritiT) [inline tests/acceptance/stepDefinitions/filesContext.js:714]: I don't think this step will be necessary.

File: tests/acceptance/features/webUIBatchAction/batchAction.feature [hunk 13]
```diff
@@ -0,0 +1,65 @@
+Feature: Visibilty of the batch actions menu
```
Comment (saw-jan) [inline tests/acceptance/features/webUIBatchAction/batchAction.feature:1]: These tests are unrelated to the PR. Please, do this in a separate PR. And then the review process will also be a lot easier.

Comment (amrita-shrestha) [reply tests/acceptance/features/webUIBatchAction/batchAction.feature]: Ok. I will do in next PR

**Review by sakshamgurung — COMMENTED** [review]
Very good!

File: tests/acceptance/features/webUIFavouritesActionMenu/favouritesActionMenu.feature (hunk 1 above)
Comment (amrita-shrestha) [reply tests/acceptance/features/webUIFavouritesActionMenu/favouritesActionMenu.feature]: I have done it in recent commit.

File: tests/acceptance/features/webUIRenameFiles/renameFilesWithoutReload.feature (hunk 2 above)
Comment (amrita-shrestha) [reply tests/acceptance/features/webUIRenameFiles/renameFilesWithoutReload.feature]: Ok, I have done it in recent commit.

**Comment by ownclouders** [comment]
Results for <strong>oCISSharingPubExpAndRoles</strong> https://drone.owncloud.com/owncloud/web/20683/66/1
The following scenarios passed on retry:
- webUISharingPublicExpire/shareByPublicLinkExpiringLinks.feature:11

File: tests/acceptance/pageObjects/FilesPageElement/contextMenu.js [hunk 14]
```diff
@@ -38,6 +39,9 @@ module.exports = {
       await this.clickMenuItem(this.ContextMenuItem.showActions)
       await this.waitForAnimationToFinish() // wait for sidebar animation to finish
       return this
+    },
+    selectrenameFile: function () {
```
Comment (saw-jan) [inline tests/acceptance/pageObjects/FilesPageElement/contextMenu.js:43]:
```suggestion
    selectRenameFile: function () {
```

File: tests/acceptance/pageObjects/FilesPageElement/filesList.js [hunk 15]
```diff
@@ -366,6 +376,16 @@ module.exports = {
...
     },
+    waitForFileRenameVisible: async function (fileName, elementType = 'any') {
+      const rowSelector = this.getFileRowSelectorByFileName(fileName, elementType)
+      let fileRename = null
+      await this.waitForFileVisible(fileName, (elementType = 'any'))
+      await this.getText('xpath', rowSelector, (result) => {
+        const res = result.value
+        fileRename = res.split('\n')
+      })
+      return fileRename[1]
```
Comment (kiranparajuli589) [inline tests/acceptance/pageObjects/FilesPageElement/filesList.js:387]: does the `result.value` always contains `\n` character?
please be cautious while doing `fileRename[1]`, the `fileRename` can still or null or it could be just a string, not an array.

File: tests/acceptance/features/webUIRenameFiles/renameFiles.feature [hunk 16]
```diff
@@ -272,3 +272,16 @@ Feature: rename files
...
+  Scenario: Rename file extension through context-menu without reload (oc10 and ocis bug demonstration)
+    When the user renames file "lorem.txt" to "lorem.md" through context-menu using the webUI
+    Then file "lorem.md" should be listed as "lore.txt" on the files-table
+    And file "lorem.md" should be listed on the sidebar
+
+  @issue-4893
+  Scenario: Rename file extension through sidebar without reload (oc10 and ocis bug demonstration)
+    When the user renames file "lorem.txt" to "lorem.md" using the webUI
+    Then file "lorem.txt" should be listed on the sidebar
+    And file "lorem.md" should be listed as "lore.txt" on the files-table
```
Comment (kiranparajuli589) [inline tests/acceptance/features/webUIRenameFiles/renameFiles.feature:286]:
```suggestion
    And file "lorem.md" should be listed as "lore.txt" on the files table
```

**Comment by phil-davis** [comment]
> Why do we need a test documenting this bug? As far as I understood our process we should write a happy path test for renaming and if that fails add it to the expected failures.

I agree - we have expected-failures in this repo.

It is OK for @amrita-shrestha to write a passing version as learning (she is new), and that will demonstrate that the underlying test code does work. (We do want to have confidence that any new/changed test step code does actually pass or fail reliably when the app works/does not work.)

But after demonstrating that, there needs to be a PR that adds a test scenario(s) the expect the correct behavior, will fail, and will be added to expected-failures.

Then a developer can work on the fix and make the test pass.

File: tests/acceptance/pageObjects/FilesPageElement/filesList.js (hunk 15 above)
Comment (amrita-shrestha) [reply tests/acceptance/pageObjects/FilesPageElement/filesList.js:387]: The result.value returns string 
```
Select file
lore.txt
699 B 10 seconds ago 
```
which is then converted into array 
` [ 'Select file', 'lore.txt', '699 B 10 seconds ago' ] `

Comment (kiranparajuli589) [reply tests/acceptance/pageObjects/FilesPageElement/filesList.js]: okay, please add some comments about the splitting

**Comment by ownclouders** [comment]
Results for <strong>oC10SharingAccept</strong> https://drone.owncloud.com/owncloud/web/20789/16/1
The following scenarios passed on retry:
- webUISharingAcceptSharesToRoot/acceptShares.feature:108

**Comment by ownclouders** [comment]
Results for <strong>oC10SharingAccept</strong> https://drone.owncloud.com/owncloud/web/20835/16/1
The following scenarios passed on retry:
- webUISharingAcceptSharesToRoot/acceptShares.feature:108

**Review by kiranparajuli589 — CHANGES_REQUESTED** [review]

File: tests/acceptance/pageObjects/FilesPageElement/filesList.js [hunk 17]
```diff
@@ -362,9 +373,35 @@ module.exports = {
...
+        }
+      })
+      if (error !== null) {
+        assert.fail(error)
+      }
+      assert.strictEqual(
+        fileName,
+        resourceName,
+        `Expected file name to be "${fileName}" but found "${resourceName}"`
+      )
```
Comment (kiranparajuli589) [inline tests/acceptance/pageObjects/FilesPageElement/filesList.js:404]: it's not a good practice to perform assertions inside the `pageObjects` method.
you can return the text found from here and assert the expected case in the context files.

Comment (amrita-shrestha) [reply tests/acceptance/pageObjects/FilesPageElement/filesList.js:404]: waitForFileVisible() is calling checkFileName() to check file name and waitForFileVisible() is called by many other functions. So, we can't return the value from checkFileName()

**Comment by phil-davis** [comment]
@amrita-shrestha other things have been merged in web today. You will need to rebase this PR and sort out any conflicts.

**Review by kiranparajuli589 — COMMENTED** [review]
other than the existing comments lgtm

**Review by phil-davis — APPROVED** [review]
LGTM

**Comment by phil-davis** [comment]
The SonarCloud coverage report shows a list of source-code files that are not touched by this PR. I have no idea why it reported this, and failed it's result.

@kulmann @pascalwengerter are you OK with us adding this test scenario?

It was in progress as a learning exercise for @amrita-shrestha before we talked about stopping nightwatch test enhancements.

**Comment by pascalwengerter** [comment]
> The SonarCloud coverage report shows a list of source-code files that are not touched by this PR. I have no idea why it reported this, and failed it's result.
> 
> @kulmann @pascalwengerter are you OK with us adding this test scenario?
> 
> It was in progress as a learning exercise for @amrita-shrestha before we talked about stopping nightwatch test enhancements.

IIRC some commits that were eventually dropped touched some actual `packages/` files, maybe SonarCloud remembers. PR in general looks good to me, happy to merge it and move forward

**Comment by phil-davis** [comment]
@kulmann please force-merge (see comments above about SonarCloud)
